File: wine.py
# Importing Required Packages
import pandas as pd
import numpy as np

# Importing Dataset
data = pd.read_csv("C:\\Data Science\\Assignments\\Module-Dimension Reduction\\Codes\\wine.csv")
data.describe()

# Removing Unnecessary columns
Data = data.drop(["Type"], axis = 1)
 
######### DIMENSION REDUCTION (PCA) ANALYSIS ######

from sklearn.decomposition import PCA
import matplotlib.pyplot as plt
from sklearn.preprocessing import scale

# Normalizing the numerical data 
data_normal = scale(Data)
data_normal

pca = PCA(n_components = 6)
pca_values = pca.fit_transform(data_normal)

# The amount of variance that each PCA explains is 
var = pca.explained_variance_ratio_
var

pca.components_
pca.components_[0]
# Cumulative variance 

var1 = np.cumsum(np.round(var, decimals = 4) * 100)
var1

# Variance plot for PCA components obtained 
plt.plot(var1, color = "red")

# PCA scores
pca_values

pca_data = pd.DataFrame(pca_values)
pca_data.columns = "comp0", "comp1", "comp2", "comp3", "comp4", "comp5"
final = pd.concat([data.Type, pca_data.iloc[:, 0:3]], axis = 1)

# Scatter diagram
import matplotlib.pylab as plt
plt.scatter(x = final.comp0, y = final.comp1)


#### HIERARCHICAL CLUSTERING ####
from scipy.spatial.distance import cdist 
from sklearn.cluster import KMeans

data = pd.read_csv("C:\\Data Science\\Assignments\\Module-Dimension Reduction\\Codes\\wine.csv")
data.describe()

Data = data.drop(["Type"], axis = 1)

############Normalizing the data#################
def norm_func(i):
    x=(i-i.min()/i.std())
    return (x)

# Normalized data frame (considering the numerical part of data)
df_norm=norm_func(Data.iloc[:,0:])
df_norm.describe()

# for creating dendrogram 
from scipy.cluster.hierarchy import linkage
import scipy.cluster.hierarchy as sch 
z=linkage(df_norm,method="complete",metric="euclidean")

# Dendrogram
plt.figure(figsize=(25,50));plt.title("Dendogram");plt.xlabel("Index");plt.ylabel("Distance")
sch.dendrogram(
    z,
    leaf_rotation=0.,
    leaf_font_size=6.,
)
plt.show()


#################NOw applying agglomerative clustering########
from sklearn.cluster import AgglomerativeClustering
h_labels=AgglomerativeClustering(n_clusters=14,affinity="euclidean",linkage="complete").fit(df_norm)
clusters_labels=pd.Series(h_labels.labels_)

data["Clusters"]=clusters_labels
df_final=data.iloc[:,[0,14,1,2,3,4,5,6,7,8,9,10,11,12,13]]
df_final.to_csv("wine_py.csv")


#### K-MEANS CLUSTERING ####

# Importing Required Packages
from sklearn.cluster import	KMeans

# Importing Dataset
data = pd.read_csv("C:\\Data Science\\Assignments\\Module-Dimension Reduction\\Codes\\wine.csv")
data.describe()

Data = data.drop(["Type"], axis = 1)

############Normalizing the data#################
def norm_func(i):
    x=(i-i.min()/i.std())
    return (x)

# Normalized data frame (considering the numerical part of data)
df_norm=norm_func(Data.iloc[:,0:])
df_norm.describe()

###### scree plot or elbow curve ############
TWSS = []
k = list(range(2, 8))

for i in k:
    kmeans = KMeans(n_clusters = i)
    kmeans.fit(df_norm)
    TWSS.append(kmeans.inertia_)
    
TWSS

# Scree plot 
plt.plot(k, TWSS, 'ro-');plt.xlabel("No_of_Clusters");plt.ylabel("total_within_SS")

# Selecting 3 clusters from the above scree plot which is the optimum number of clusters 
model = KMeans(n_clusters = 3)
model.fit(df_norm)

model.labels_ # getting the labels of clusters assigned to each row 
mb = pd.Series(model.labels_)  # converting numpy array into pandas series object 
Data['clust'] = mb # creating a  new column and assigning it to new column 

Data.head()
df_norm.head()

Wine = Data.iloc[:,[0,13,1,2,3,4,5,6,7,8,9,10,11,12]]
Wine.head()

Wine.iloc[:, 1:14].groupby(Data.clust).mean()

Data.to_csv("Kmeans_Wine.csv", encoding = "utf-8")

import os
os.getcwd()

File: wine.R
# Loading data
input <- read.csv(file.choose())
View(input)


## Removing Unnecessary columns
data <- input[, -1]
attach(data)

# Exploratory data analysis:
# 1. Measures of central tendency
# 2. Measures of dispersion
# 3. Third moment business decision
# 4. Fourth moment business decision
# 5. Probability distributions of variables 
# 6. Graphical representations (Histogram, Box plot, Dot plot, Stem & Leaf plot, Bar plot, etc.)

summary(data)


# Graphical representations (Histogram, Box plot, Dot plot, Stem & Leaf plot, Bar plot, etc.)

# Box plot Representation

boxplot(Alcohol, col = "dodgerblue4",main = "Alcohol")
boxplot(Malic, col = "dodgerblue4",main = "Malic")
boxplot(Alcalinity, col = "dodgerblue4",main = "Alcalinity")
boxplot(Magnesium, col = "dodgerblue4",main = "Magnesium")
boxplot(Phenols, col = "dodgerblue4",main = "Phenols")
boxplot(Proanthocyanins, col = "dodgerblue4",main = "Proanthocyanins")


# Histogram Representation

hist(Alcohol, col = "blue",main = "Alcohol")
hist(Malic, col = "blue",main = "Malic")
hist(Alcalinity, col = "blue",main = "Alcalinity")
hist(Magnesium, col = "blue",main = "Magnesium")
hist(Phenols, col = "blue",main = "Phenols")
hist(Proanthocyanins, col = "blue",main = "Proanthocyanins")


########## DIMENSION REDUCTION (PCA) ANALYSIS #############

pcaObj <- princomp(data, cor = TRUE, scores = TRUE, covmat = NULL)

str(pcaObj)
summary(pcaObj)

loadings(pcaObj)

plot(pcaObj) # graph showing importance of principal components 

biplot(pcaObj)

plot(cumsum(pcaObj$sdev * pcaObj$sdev) * 100 / (sum(pcaObj$sdev * pcaObj$sdev)), type = "b")

pcaObj$scores
pcaObj$scores[, 1:3]

# Top 3 pca scores 
final <- cbind(input[, 1], pcaObj$scores[, 1:3])
View(final)


# Scatter diagram
plot(final)

#### HIERARCHICAL CLUSTERING ####

# Loading the dataset
input <- read.csv(file.choose())
View(input)


# Removing Unnecessary columns
Data <- input[, -1]
attach(Data)

summary(Data)

# Normalize the data
normalized_data <- scale(Data[, 1:13]) 

summary(normalized_data)

# Distance matrix
d <- dist(normalized_data, method = "euclidean") 

fit <- hclust(d, method = "ward.D2")

# Display dendrogram
plot(fit) 
plot(fit, hang = -1)

groups <- cutree(fit, k =14)# Cut tree into 14 clusters

rect.hclust(fit, k =14, border = "red")

cluster <- as.matrix(groups)

final <- data.frame(cluster, Data)

aggregate(Data[, 1:11], by = list(final$cluster), FUN = mean)

library(readr)
write_csv(final, "Wine_hierarachial_R.csv")

getwd()

#### K MEANS CLUSTERING ####

# Importing the dataset
input <- read.csv(file.choose())
View(input)


## Removing Unnecessary Columns
Data <- input[, -1]
attach(Data)
str(Data)

summary(Data)

# Normalize the data
normalized_data <- scale(Data[, 1:13]) # As we already removed "Type" column so all columns need to normalize

summary(normalized_data)

# Elbow curve to decide the k value
twss <- NULL
for (i in 2:12) {
  twss <- c(twss, kmeans(normalized_data, centers = i)$tot.withinss)
}
twss

# Look for an "elbow" in the scree plot
plot(2:12, twss, type = "b", xlab = "Number of Clusters", ylab = "Within groups sum of squares")
title(sub = "K-Means Clustering Scree-Plot")


# 4 Cluster Solution
fit <- kmeans(normalized_data, 4) 
str(fit)
fit$cluster
final <- data.frame(fit$cluster, Data) # Append cluster membership

A <- aggregate(Data[, 1:13], by = list(fit$cluster), FUN = mean)

library(readr)
write_csv(A, "Wine_kmeans_R.csv")

getwd()
